icu4c/source/common/uniquecharstr.h

```diff
@@ -48,21 +50,41 @@ class UniqueCharStrings {
 
     /**
      * Adds a string and returns a unique number for it.
-     * The string's buffer contents must not change, nor move around in memory,
-     * while this UniqueCharStrings is in use.
-     * The string contents must be NUL-terminated exactly at s.length().
-     *
+     * If copy is false, then the string's buffer contents must not change,
+     * nor move around in memory, while this UniqueCharStrings is in use.
```

Comment:
Suggestion: `UnicodeString` itself has a constructor that points to other memory:

https://unicode-org.github.io/icu-docs/apidoc/dev/icu4c/classicu_1_1UnicodeString.html#a28a13b2ef3a75b8a947bcd040a98e206

So, maybe just change this function to take the UnicodeString by value, and call sites that point to constant in-memory data can use the readonly-alias UnicodeString constructor.